### PR TITLE
Midpoints+Klein coordinates

### PR DESCRIPTION
**Pull request by rrkarim**
Added Klein coordinates and midpoint. Though the naming of functions should be reconsidered. Also don't know how to test `midpoint` method.

### CODE REVIEW

File: geoopt/manifolds/poincare/math.py [hunk 1]
```diff
@@ -1373,7 +1373,9 @@ def midpoint(x, c=1.0, dim=-1):
 def _midpoint(x, c=1.0, dim: int = -1):
     x = _poincare2klein(x, c)
```
Comment (ferrine) [inline geoopt/manifolds/poincare/math.py]: You forget dim here

File: geoopt/manifolds/poincare/math.py [hunk 2]
```diff
@@ -1345,3 +1345,90 @@ def egrad2rgrad(x, grad, *, c=1.0, dim=-1):
...
+    tensor
+        midpoint
+    """
+    return _mean(x, c, dim=dim)
+
+
+def _midpoint(x, c=1.0, dim: int = -1):
+    x = _poincare2klein(x, c)
+    lambdas = lambda_x(x, c=c, keepdim=True)
+    mean = torch.sum(lambdas * x, dim=dim, keepdim=True) / torch.sum(
```
Comment (ferrine) [inline geoopt/manifolds/poincare/math.py]: Keepdim should be used here

File: geoopt/manifolds/poincare/math.py [hunk 3]
```diff
@@ -1345,3 +1345,90 @@ def egrad2rgrad(x, grad, *, c=1.0, dim=-1):
...
+        lambdas, dim=dim, keepdim=True
+    )
+    mean = _klein2poincare(mean, c)
+    return mean.squeeze(dim)
+
+
+def poincare2klein(x, c=1.0, dim=-1):
```
Comment (ferrine) [inline geoopt/manifolds/poincare/math.py]: C and dim should be keyword only

File: tests/test_poincare_math.py [hunk 4]
```diff
@@ -370,3 +370,8 @@ def test_distance2plane(a, c):
     dist = poincare.math.dist2plane(z, a, vr, c=c)
 
     np.testing.assert_allclose(dist, dist1, atol=1e-5, rtol=1e-5)
+
+def test_klein_transition(a, c):
```
Comment (ferrine) [inline tests/test_poincare_math.py]: Do you know any test for midpoint?

File: geoopt/manifolds/poincare/math.py [hunk 5]
```diff
@@ -1345,3 +1345,90 @@ def egrad2rgrad(x, grad, *, c=1.0, dim=-1):
...
+        ball negative curvature
+    dim : int
+        reduction dimension for operations
+
+    Returns
+    -------
+    tensor
+        midpoint
+    """
+    return _mean(x, c, dim=dim)
```
Comment (ferrine) [inline geoopt/manifolds/poincare/math.py]: Is it a typo?

File: geoopt/manifolds/poincare/math.py (hunk 2 above)
Comment (CoderINusE) [reply geoopt/manifolds/poincare/math.py]: as an argument for `midpoint`?

File: tests/test_poincare_math.py (hunk 4 above)
Comment (CoderINusE) [reply tests/test_poincare_math.py]: I wasn't able to think of one. I can check with zero curvature.

File: tests/test_poincare_math.py [hunk 6]
```diff
@@ -370,3 +370,17 @@ def test_distance2plane(a, c):
...
+
+
+def test_klein_transition(a, c):
+    kl = poincare.math.poincare2klein(a, c=c)
+    b = poincare.math.klein2poincare(kl, c=c)
+    np.testing.assert_allclose(a, b, atol=1e-6, rtol=1e-6)
+
+
+def test_midpoint(a, c):
+    # test the midpoint with zero curvature
```
Comment (CoderINusE) [inline tests/test_poincare_math.py]: @ferrine this test doesn't pass. `poincare.math.midpoint` output is `mids_b = 10 * mids_a`.

File: tests/test_poincare_math.py [hunk 7]
```diff
@@ -370,3 +370,17 @@ def test_distance2plane(a, c):
...
+    mids_a = torch.mean(a, dim=-1)
+    mids_b = poincare.math.midpoint(a, c=0, dim=-1)
+    np.testing.assert_allclose(mids_a, mids_b, atol=1e-6, rtol=1e-6)
```
Comment (CoderINusE) [inline tests/test_poincare_math.py]: @ferrine this test doesn't pass. `poincare.math.midpoint` output is `mids_b = 10 * mids_a`.

Comment (newkozlukov) [reply tests/test_poincare_math.py]: Consider also adding a test that barycentre of a measure concentrated on two points lies on a geodesic connecting these points

Comment (ferrine) [reply tests/test_poincare_math.py]: So far there are concerns that they indeed lie there

Comment (ferrine) [reply tests/test_poincare_math.py]: I've checked this for poincare ball of dimension 2 ant it was all right, intuition breaks for dim>2 as in test

Comment (ferrine) [reply tests/test_poincare_math.py]: Remark for above. Confirmed for dim>2 and 2 points it is ok once points are not close to the boundary. From more than 2 points running and optimization to find barycenter ($argmin_y \mathbb{E}d(y, x)^2$) gives different result than poincare_mean. Checking loss values for these 2 points there is a confidence that optimization has converged and found the barycenter.
